providers/implementations/keymgmt/ml_kem_kmgmt.c

```diff
@@ -0,0 +1,631 @@
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_PRIV_KEY, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    return arr;
+}
+
+/*
+ * It is assumed the key is guaranteed non-NULL here, and is from this provider
+ */
+static int ml_kem_get_params(void *vkey, OSSL_PARAM params[])
+{
+    ML_KEM_PROVIDER_KEYPAIR *key = vkey;
+    vinfo_t v = key->vinfo;
+    OSSL_PARAM *p;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_BITS);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, v->bits))
+            return 0;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_SECURITY_BITS);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, v->secbits))
+            return 0;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_MAX_SIZE);
+    if (p != NULL)
+        if (!OSSL_PARAM_set_int(p, v->ctext_bytes))
+            return 0;
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY);
+    if (p != NULL && have_keys(key)) {
+        if (p->data_type != OSSL_PARAM_OCTET_STRING)
+            return 0;
+        p->return_size = key->vinfo->pubkey_bytes;
+        if (p->data != NULL) {
+            if (p->data_size < p->return_size)
+                return 0;
+            if (!ossl_ml_kem_vencode_public_key(v, p->data, p->return_size,
+                                                key->pubkey, key->prvkey))
+                return 0;
+        }
+    }
+
+    p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_PRIV_KEY);
+    if (p != NULL && key->prvkey != NULL) {
+        if (p->data_type != OSSL_PARAM_OCTET_STRING)
+            return 0;
+        p->return_size = key->vinfo->prvkey_bytes;
+        if (p->data != NULL) {
+            if (p->data_size < p->return_size)
+                return 0;
+            if (!ossl_ml_kem_vencode_private_key(v, p->data, p->return_size,
+                                                 key->prvkey))
+                return 0;
+        }
+    }
+    return 1;
+}
+
+static const OSSL_PARAM *ml_kem_settable_params(void *provctx)
+{
+    static const OSSL_PARAM arr[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    return arr;
+}
+
+static int ml_kem_set_params(void *vkey, const OSSL_PARAM params[])
+{
+    ML_KEM_PROVIDER_KEYPAIR *key = vkey;
+    vinfo_t v = key->vinfo;
+    const OSSL_PARAM *p;
+    const void *pubenc = NULL;
+    size_t publen = 0;
+
+    if (params == NULL)
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY);
+    if (p == NULL)
+        return 1;
+
+    /* Key mutation is reported generally not allowed */
+    if (have_keys(key)) {
+        ERR_raise_data(ERR_LIB_PROV,
+                       PROV_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE,
+                       "ML-KEM keys cannot be mutated");
+        return 0;
+    }
+    /* An unlikely failure mode is the parameter having some unexpected type */
+    if (!OSSL_PARAM_get_octet_string_ptr(p, &pubenc, &publen))
```

Comment:
There are not too many places that use get_octet_string_ptr(). Are you sure that is what you want to do here..